components/proxy/src/test/java/com/hotels/styx/routing/handlers/PathPrefixRouterTest.java

```diff
@@ -0,0 +1,134 @@
+        when(Routingcontext.refLookup()).thenReturn(ref -> routingObjects.get(ref));
+    }
+
+    @Test
+    public void no_routes_always_throws_NoServiceConfiguredException() throws Exception {
+        PathPrefixRouter router = buildRouter(emptyMap());
+
+        assertThrows(NoServiceConfiguredException.class, () ->
+                Mono.from(router.handle(LiveHttpRequest.get("/").build(), null)).block()
+        );
+    }
+
+    @Test
+    public void root_handler_handles_sub_paths() throws Exception {
+        RoutingObject rootHandler = mock(RoutingObject.class);
+
+        routingObjects.put(new StyxObjectReference("rootHandler"), rootHandler);
+
+        PathPrefixRouter router = buildRouter(singletonMap("/", "rootHandler"));
+
+        testRequestRoute(router, "/", rootHandler);
+        testRequestRoute(router, "/foo/bar", rootHandler);
+    }
+
+    @Test
+    public void most_specific_path_is_chosen() throws Exception {
+        RoutingObject rootHandler = mock(RoutingObject.class);
+        RoutingObject fooFileHandler = mock(RoutingObject.class);
+        RoutingObject fooPathHandler = mock(RoutingObject.class);
+        RoutingObject fooBarFileHandler = mock(RoutingObject.class);
+        RoutingObject fooBarPathHandler = mock(RoutingObject.class);
+        RoutingObject fooBazFileHandler = mock(RoutingObject.class);
+
+        routingObjects.put(new StyxObjectReference("rootHandler"), rootHandler);
+        routingObjects.put(new StyxObjectReference("fooFileHandler"), fooFileHandler);
+        routingObjects.put(new StyxObjectReference("fooPathHandler"), fooPathHandler);
+        routingObjects.put(new StyxObjectReference("fooBarFileHandler"), fooBarFileHandler);
+        routingObjects.put(new StyxObjectReference("fooBarPathHandler"), fooBarPathHandler);
+        routingObjects.put(new StyxObjectReference("fooBazFileHandler"), fooBazFileHandler);
+
+        PathPrefixRouter router = buildRouter(ImmutableMap.<String, String>builder()
+                .put("/", "rootHandler")
+                .put("/foo", "fooFileHandler")
+                .put("/foo/", "fooPathHandler")
+                .put("/foo/bar", "fooBarFileHandler")
+                .put("/foo/bar/", "fooBarPathHandler")
+                .put("/foo/baz", "fooBazFileHandler")
+                .build());
+
+        testRequestRoute(router, "/", rootHandler);
+        testRequestRoute(router, "/foo", fooFileHandler);
+        testRequestRoute(router, "/foo/x", fooPathHandler);
+        testRequestRoute(router, "/foo/", fooPathHandler);
+        testRequestRoute(router, "/foo/bar", fooBarFileHandler);
+        testRequestRoute(router, "/foo/bar/x", fooBarPathHandler);
+        testRequestRoute(router, "/foo/bar/", fooBarPathHandler);
+        testRequestRoute(router, "/foo/baz/", fooBazFileHandler);
+        testRequestRoute(router, "/foo/baz/y", fooBazFileHandler);
+    }
+
+    private PathPrefixRouter buildRouter(Map<String, String> prefixRoutes) {
```

Comment:
Make `PrefixRoute` package private. Then you can simply invoke object's constructor, instead of having to go through the factory.

Reply:
But this won't work because `Builtins.build` uses a lambda for an object reference lookup:

```
    public static RoutingObject build(...)
        ...
        } else if (configNode instanceof StyxObjectReference) {
            return (request, httpContext) -> context.refLookup()
                    .apply((StyxObjectReference) configNode)
                    .handle(request, httpContext);
        ...
```

At the moment this lambda could be promoted to a separate class. Perhaps like so:

```
public class ObjectRefLookup implements RoutingObject {
    private final RouteRefLookup routeRefLookup;
    private final StyxObjectReference configNode;

    public ObjectRefLookup(RouteRefLookup routeRefLookup, StyxObjectReference configNode) {
        this.routeRefLookup = routeRefLookup;
        this.configNode = configNode;
    }

    @Override
    public Eventual<LiveHttpResponse> handle(LiveHttpRequest request, HttpInterceptor.Context context) {
        return routeRefLookup.apply(configNode)
                .handle(request, context);
    }
}
```

But sadly this class would have to have a public visibility due to our packaging layout.